File: goldencheck/engine/history.py
"""Scan history — append-only JSONL log of scan results."""
from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

from goldencheck.models.finding import Finding, Severity
from goldencheck.models.profile import DatasetProfile

logger = logging.getLogger(__name__)

__all__ = ["record_scan", "load_history", "get_previous_scan", "ScanRecord"]

HISTORY_DIR = Path(".goldencheck")
HISTORY_FILE = HISTORY_DIR / "history.jsonl"


@dataclass
class ScanRecord:
    timestamp: str
    file: str
    rows: int
    columns: int
    grade: str
    score: int
    errors: int
    warnings: int
    findings_count: int


def record_scan(
    file: str | Path,
    profile: DatasetProfile,
    findings: list[Finding],
) -> None:
    """Append a scan record to the history log."""
    errors = sum(1 for f in findings if f.severity == Severity.ERROR)
    warnings = sum(1 for f in findings if f.severity == Severity.WARNING)

    by_col: dict[str, dict[str, int]] = {}
    for f in findings:
        if f.severity >= Severity.WARNING:
            by_col.setdefault(f.column, {"errors": 0, "warnings": 0})
            key = "errors" if f.severity == Severity.ERROR else "warnings"
            by_col[f.column][key] = by_col[f.column].get(key, 0) + 1

    grade, score = profile.health_score(findings_by_column=by_col)

    record = {
        "timestamp": datetime.now().isoformat(timespec="seconds"),
        "file": str(Path(file).name),
        "rows": profile.row_count,
        "columns": profile.column_count,
        "grade": grade,
        "score": score,
        "errors": errors,
        "warnings": warnings,
        "findings_count": len(findings),
    }

    try:
        HISTORY_DIR.mkdir(parents=True, exist_ok=True)
        with open(HISTORY_FILE, "a") as f:
            f.write(json.dumps(record) + "\n")
    except OSError as e:
        logger.warning("Failed to write scan history: %s", e)


def load_history(
    file_filter: str | None = None,
    last_n: int | None = None,
) -> list[ScanRecord]:
    """Load scan records from history."""
    if not HISTORY_FILE.exists():
        return []

    records = []
    with open(HISTORY_FILE) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                data = json.loads(line)
                record = ScanRecord(**data)
                if file_filter and record.file != file_filter:
                    continue
                records.append(record)
            except (json.JSONDecodeError, TypeError):
                continue

    if last_n:
        records = records[-last_n:]

    return records


def get_previous_scan(file: str | Path) -> ScanRecord | None:
    """Get the most recent scan record for a file."""
    name = str(Path(file).name)
    records = load_history(file_filter=name)
    return records[-1] if records else None

File: packages/goldencheck-js/src/core/engine/history.ts
/**
 * Scan history — append-only JSONL log of scan results.
 * Port of goldencheck/engine/history.py.
 * Types are edge-safe; read/write functions require Node.js (node:fs).
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from "node:fs";
import { resolve, join, dirname } from "node:path";

import type { Finding, DatasetProfile } from "../types.js";
import { Severity, healthScore } from "../types.js";

// --- Types ---

export interface ScanRecord {
  readonly timestamp: string; // ISO 8601
  readonly file: string;
  readonly rows: number;
  readonly columns: number;
  readonly grade: string;
  readonly score: number;
  readonly errors: number;
  readonly warnings: number;
  readonly findingsCount: number;
}

// --- Constants ---

const HISTORY_DIR = ".goldencheck";
const HISTORY_FILE = join(HISTORY_DIR, "history.jsonl");

// --- Public API ---

/**
 * Append a scan record to `.goldencheck/history.jsonl`.
 * Computes health grade from findings using per-column capped scoring.
 */
export function recordScan(
  file: string,
  profile: DatasetProfile,
  findings: readonly Finding[],
): void {
  const errors = findings.filter((f) => f.severity === Severity.ERROR).length;
  const warnings = findings.filter((f) => f.severity === Severity.WARNING).length;

  // Build per-column breakdown for capped health score
  const byCol: Record<string, { errors: number; warnings: number }> = {};
  for (const f of findings) {
    if (f.severity >= Severity.WARNING) {
      if (!byCol[f.column]) {
        byCol[f.column] = { errors: 0, warnings: 0 };
      }
      if (f.severity === Severity.ERROR) {
        byCol[f.column]!.errors += 1;
      } else {
        byCol[f.column]!.warnings += 1;
      }
    }
  }

  const hs = healthScore(byCol);

  const record: Record<string, unknown> = {
    timestamp: new Date().toISOString().replace(/\.\d{3}Z$/, ""),
    file: resolve(file),
    rows: profile.rowCount,
    columns: profile.columnCount,
    grade: hs.grade,
    score: hs.points,
    errors,
    warnings,
    findingsCount: findings.length,
  };

  try {
    mkdirSync(HISTORY_DIR, { recursive: true });
    writeFileSync(HISTORY_FILE, JSON.stringify(record) + "\n", { flag: "a" });
  } catch (e) {
    console.warn("Failed to write scan history:", e);
  }
}

/**
 * Load scan records from history.
 * Optionally filter by file path and limit to the last N records.
 */
export function loadHistory(fileFilter?: string, lastN?: number): ScanRecord[] {
  if (!existsSync(HISTORY_FILE)) {
    return [];
  }

  let content: string;
  try {
    content = readFileSync(HISTORY_FILE, "utf-8");
  } catch {
    return [];
  }

  const records: ScanRecord[] = [];
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;

    try {
      const data = JSON.parse(trimmed) as Record<string, unknown>;
      const record: ScanRecord = {
        timestamp: String(data.timestamp ?? ""),
        file: String(data.file ?? ""),
        rows: Number(data.rows ?? 0),
        columns: Number(data.columns ?? 0),
        grade: String(data.grade ?? ""),
        score: Number(data.score ?? 0),
        errors: Number(data.errors ?? 0),
        warnings: Number(data.warnings ?? 0),
        findingsCount: Number(data.findingsCount ?? data.findings_count ?? 0),
      };

      if (fileFilter && record.file !== fileFilter) {
        continue;
      }
      records.push(record);
    } catch {
      // Skip malformed lines
      continue;
    }
  }

  if (lastN && lastN > 0) {
    return records.slice(-lastN);
  }

  return records;
}

/**
 * Get the most recent scan record for a given file.
 * Returns null if no previous scan exists.
 */
export function getPreviousScan(file: string): ScanRecord | null {
  const resolved = resolve(file);
  const records = loadHistory(resolved);
  return records.length > 0 ? records[records.length - 1]! : null;
}
